vault/external_tests/raft/raft_autopilot_test.go

```diff
@@ -346,6 +350,226 @@ func TestRaft_AutoPilot_Peersets_Equivalent(t *testing.T) {
+			for i := 1; i < len(cluster.Cores); i++ {
+				core := cluster.Cores[i]
+				core.UnderlyingRawStorage.(*raft.RaftBackend).SetServerAddressProvider(addressProvider)
+				// Only add the number of nodes specified as 'voters' the rest won't be.
+				if voters < tc.numVoters {
+					joinAsVoterAndUnseal(t, core, cluster)
+				} else {
+					joinAsNonVoterAndUnseal(t, core, cluster)
+				}
+				voters++
+			}
+			testhelpers.WaitForActiveNodeAndStandbys(t, cluster)
+
+			// Actual code related to testing the return values of the KnownServers func
+			backend := leader.GetRaftBackend()
+			d := raft.NewDelegate(backend)
+			servers := d.KnownServers()
+			voters = 0
+			for id, srv := range servers {
+				fmt.Printf("known servers: id: %v, node type: %v\n", id, srv.NodeType)
+				if srv.NodeType == "voter" {
+					voters++
+				}
+			}
+			require.Equal(t, tc.numVoters, voters, "known servers reported a number of voters that wasn't expected")
+		})
+	}
+}
+
+// TestRaft_Autopilot_DeadServerCleanup tests that dead servers are correctly removed by Vault and autopilot when a node stops and a replacement node joins.
+// The expected behavior is that removing a node from a 3 node cluster wouldn't remove it from Raft until a replacement voter had joined and stabilized/been promoted.
+func TestRaft_Autopilot_DeadServerCleanup(t *testing.T) {
+	// Some involved setup, so we can control the nodes being added to the cluster and have access to the leader backend.
+	conf, opts := teststorage.ClusterSetup(nil, nil, teststorage.RaftBackendSetup)
+	conf.DisableAutopilot = false
+	opts.NumCores = 4
+	opts.SetupFunc = nil
+	opts.PhysicalFactoryConfig = map[string]interface{}{
+		"autopilot_reconcile_interval": "300ms",
+		"autopilot_update_interval":    "100ms",
+	}
+
+	cluster := vault.NewTestCluster(t, conf, opts)
+	cluster.Start()
+	defer cluster.Cleanup()
+	leader, addressProvider := setupLeaderAndUnseal(t, cluster)
+
+	// Join 2 extra nodes manually, store the 3rd for later
+	core1 := cluster.Cores[1]
+	core2 := cluster.Cores[2]
+	core3 := cluster.Cores[3]
+	core1.UnderlyingRawStorage.(*raft.RaftBackend).SetServerAddressProvider(addressProvider)
+	core2.UnderlyingRawStorage.(*raft.RaftBackend).SetServerAddressProvider(addressProvider)
+	core3.UnderlyingRawStorage.(*raft.RaftBackend).SetServerAddressProvider(addressProvider)
+	joinAsVoterAndUnseal(t, core1, cluster)
+	joinAsVoterAndUnseal(t, core2, cluster)
+	core3, cluster.Cores = cluster.Cores[len(cluster.Cores)-1], cluster.Cores[:len(cluster.Cores)-1]
+	testhelpers.WaitForActiveNodeAndStandbys(t, cluster)
+
+	// Wait for servers to sort themselves out...
+	timeoutGrace := 2 * time.Second
+	config, err := leader.Client.Sys().RaftAutopilotConfiguration()
+	deadline := time.Now().Add(config.ServerStabilizationTime).Add(timeoutGrace)
+	healthy := false
+	for time.Now().Before(deadline) {
+		state, err := leader.Client.Sys().RaftAutopilotState()
+		require.NoError(t, err)
+		if state.Healthy {
+			healthy = true
```

Comment:
we should also break out of the loop here, so we don't have to wait the entire deadline

Reply:
I think it needs to wait for the stabilisation time to pass in order to be able to be promoted etc. I've tuned the settings to 5s so it won't have to wait long. Good catch though, thank you.